### PR TITLE
[FLINK-4309] Potential null pointer dereference in DelegatingCo…

### PR DESCRIPTION
**Pull request by tsunny**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

…nfiguration#keySet()

### CODE REVIEW

File: flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java [hunk 1]
```diff
@@ -88,4 +89,51 @@ private String typeParamToString(Class<?>[] classes) {
...
+	public void testDelegationConfigurationWithNullPrefix() {
+
+		Configuration backingConf = new Configuration();
+		backingConf.setValueInternal("test-key", "value");
+
+		DelegatingConfiguration configuration = new DelegatingConfiguration(
+				backingConf, null);
+		Set<String> keySet = configuration.keySet();
+
+		assertTrue(keySet.equals(backingConf.keySet()));
```
Comment (wuchong) [inline flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java]: minor, it's better to use `assertEquals(keySet, backingConf.keySet())`  here I think. Below is the same

**Comment by wuchong** [comment]
LGTM +1

File: flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java [hunk 2]
```diff
@@ -88,4 +90,49 @@ private String typeParamToString(Class<?>[] classes) {
...
+	@Test
+	public void testDelegationConfigurationWithPrefix() {
+
+		String prefix = "pref-";
+
+		/*
+		 * Key matches the prefix
+		 */
+		Configuration backingConf = new Configuration();
+		backingConf.setValueInternal("pref-key", "value");
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java]: I think `"pref-"` can be replaced by the `prefix` variable, just to make it consistent.

File: flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java [hunk 3]
```diff
@@ -178,14 +178,19 @@ public String toString() {
...
 	@Override
 	public Set<String> keySet() {
+
 		final HashSet<String> set = new HashSet<String>();
-		final int prefixLen = this.prefix == null ? 0 : this.prefix.length();
 
 		for (String key : this.backingConfig.keySet()) {
-			if (key.startsWith(this.prefix)) {
+
+			if (this.prefix == null) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java]: Can we move the condition out of the loop? The condition does not change over different iteration so we don't have to perform it every time.

File: flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java [hunk 4]
```diff
@@ -178,14 +178,19 @@ public String toString() {
 
 	@Override
 	public Set<String> keySet() {
+
```
Comment (uce) [inline flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java]: We usually don't have an empty line when a method block starts (also applies to lines 185 and in `DelegatingConfigurationTest` lines 96 and 110).

File: flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java [hunk 5]
```diff
@@ -88,4 +90,49 @@ private String typeParamToString(Class<?>[] classes) {
...
+
+		/*
+		 * Key matches the prefix
+		 */
+		Configuration backingConf = new Configuration();
+		backingConf.setValueInternal(prefix + expectedKey, "value");
+
+		DelegatingConfiguration configuration = new DelegatingConfiguration(backingConf, prefix);
+		Set<String> keySet = configuration.keySet();
+
```
Comment (uce) [inline flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java]: empty line

File: flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java [hunk 6]
```diff
@@ -88,4 +90,49 @@ private String typeParamToString(Class<?>[] classes) {
...
+		Configuration backingConf = new Configuration();
+		backingConf.setValueInternal("test-key", "value");
+
+		DelegatingConfiguration configuration = new DelegatingConfiguration(
+				backingConf, null);
+		Set<String> keySet = configuration.keySet();
+
+		assertEquals(keySet, backingConf.keySet());
+
```
Comment (uce) [inline flink-core/src/test/java/org/apache/flink/configuration/DelegatingConfigurationTest.java]: empty line

**Comment by tillrohrmann** [comment]
Thanks for your contribution @tsunny. Your changes look good. I only had a minor comment.

File: flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java (hunk 3 above)
Comment (uce) [inline flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java]: Yes, we can use can use `addAll`

**Comment by uce** [comment]
Changes look good all in all. Could you also add a `org.apache.flink.util.Preconditions.checkNotNull` check for the backing config in the constructor?

File: flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java (hunk 3 above)
Comment (uce) [inline flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java]: Actually, it's probably better to just return the keySet of the backing config.

File: flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java [hunk 7]
```diff
@@ -56,6 +57,7 @@ public DelegatingConfiguration() {
 	 */
 	public DelegatingConfiguration(Configuration backingConfig, String prefix)
 	{
+		Preconditions.checkNotNull(backingConfig);
 		this.backingConfig = backingConfig;
```
Comment (uce) [inline flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java]: Trivial note: We can also write `this.backingConfig = checkNotNull(backingConfig)` here.

**Comment by uce** [comment]
Thanks for addressing our comments! I'm going to merge this later today.

**Comment by tsunny** [comment]
No problem. Happy to contribute. Thanks for the review comments.

**Pull request closed without merging**